src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Integer/DenseIntegralFrozenDictionary.cs

```diff
@@ -0,0 +1,234 @@
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This is dialable. The closer the value gets to 0, the more likely this implementation will be used,
+        /// and the more memory will be consumed to store the values. The value of 0.1 means that up to 90% of the
+        /// slots in the values array may be unused.
+        /// </para>
+        /// <para>
+        /// As an example, DaysOfWeek's min is 0, its max is 6, and it has 7 values, such that 7 / (6 - 0 + 1) = 1.0; thus
+        /// with a threshold of 0.1, DaysOfWeek will use this implementation. But SocketError's min is -1, its max is 11004, and
+        /// it has 47 values, such that 47 / (11004 - (-1) + 1) = 0.004; thus, SocketError will not use this implementation.
+        /// </para>
+        /// </remarks>
+        private const double CountToLengthRatio = 0.1;
+
+        public static bool TryCreate<TKey, TValue>(Dictionary<TKey, TValue> source, [NotNullWhen(true)] out FrozenDictionary<TKey, TValue>? result)
+            where TKey : notnull
+        {
+            // Int32 and integer types that fit within Int32. This is to minimize difficulty later validating that
+            // inputs are in range of int: we can always cast everything to Int32 without loss of information.
+
+            if (typeof(TKey) == typeof(byte) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(byte)))
+                return TryCreate<TKey, byte, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(sbyte) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(sbyte)))
+                return TryCreate<TKey, sbyte, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(ushort) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(ushort)))
+                return TryCreate<TKey, ushort, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(short) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(short)))
+                return TryCreate<TKey, short, TValue>(source, out result);
+
+            if (typeof(TKey) == typeof(int) || (typeof(TKey).IsEnum && typeof(TKey).GetEnumUnderlyingType() == typeof(int)))
+                return TryCreate<TKey, int, TValue>(source, out result);
+
+            result = null;
+            return false;
+        }
+
+        private static bool TryCreate<TKey, TKeyUnderlying, TValue>(Dictionary<TKey, TValue> source, [NotNullWhen(true)] out FrozenDictionary<TKey, TValue>? result)
+            where TKey : notnull
+            where TKeyUnderlying : unmanaged, IBinaryInteger<TKeyUnderlying>
+        {
+            // Start enumerating the dictionary to ensure it has at least one element.
+            Dictionary<TKey, TValue>.Enumerator e = source.GetEnumerator();
+            if (e.MoveNext())
```

Comment:
Empty sources should have been filtered out by caller (`EmptyFrozenDictionary` is used for empty sources) so the first `e.MoveNext()` should always return true. Then the code block can reduce an indent.

Reply:
I still wanted the check in case the Dictionary erroneously changed concurrently. An assert would still be hittable in such a case. I'm not concerned about an extra layer of indentation.